src/server/req_delete.c

```diff
@@ -236,30 +236,68 @@ check_deletehistoryjob(struct batch_request * preq)
 		}
 	}
 
-	historyjob = svr_chk_histjob(histpjob);
-	if (historyjob == PBSE_HISTJOBID) {
+	if (svr_chk_histjob(pjob) != PBSE_HISTJOBID) {
+		/*
+		 *  Job is in a Non Finished state . It must be terminated and then its history
+		 * should be purged .
+		 */
+		return FALSE;
+	}
+
+	/* Issue history job delete request to remote server if job is moved. */
+	if (pjob->ji_qs.ji_state == JOB_STATE_MOVED)
+		issue_delete(pjob);
+
+	if (jt == IS_ARRAY_NO) {
 		snprintf(log_buffer, sizeof(log_buffer),
 			msg_job_history_delete, preq->rq_user,
 			preq->rq_host);
 		log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_JOB, LOG_INFO,
 			jid,
 			log_buffer);
 
-		/* Issue history job delete request to remote server if job is moved. */
-		if (histpjob->ji_qs.ji_state == JOB_STATE_MOVED)
-			issue_delete(histpjob);
+		job_purge(pjob);
+		preq->rq_reply.brp_code = PBSE_HISTJOBDELETED;
+		reply_send(preq);
+		return TRUE;
+	} else if (jt == IS_ARRAY_ArrayJob) {
+
+		/* The Array Job itself ... */
+		/* for each subjob of array do purge */
+
+		if (pjob->ji_ajtrk) {
+			for (i = 0; i < pjob->ji_ajtrk->tkm_ct; i++) {
+				sjid = mk_subjob_id(pjob, i);
 
-		job_purge(histpjob);
+				psjob = find_job(sjid);
```

Comment:
I think we no need to do find_job() for all subjobs. For example we can avoid find_job() for subjobs that are in 'Q' state.

Reply:
I doubt if we really encounter a history array job containing a queued subjob. HoweverI have fixed it in rc 25 https://github.com/PBSPro/pbspro/pull/590/commits/f5c0e45e72a2ec7ef7af6513cb969adeeb697a2f